README.md

[![badge](https://img.shields.io/twitter/follow/api_video?style=social)](https://twitter.com/intent/follow?screen_name=api_video)
&nbsp; [![badge](https://img.shields.io/github/stars/apivideo/private.a.video?style=social)](https://github.com/apivideo/private.a.video)
&nbsp; [![badge](https://img.shields.io/discourse/topics?server=https%3A%2F%2Fcommunity.api.video)](https://community.api.video)
![](https://github.com/apivideo/.github/blob/main/assets/apivideo_banner.png)
<h1 align="center">record.a.video</h1>

[api.video](https://api.video) is the video infrastructure for product builders. Lightning fast
video APIs for integrating, scaling, and managing on-demand & low latency live streaming features in
your app.

# Table of contents

- [Table of contents](#table-of-contents)
- [Project description](#project-description)
- [Getting started](#getting-started)
  - [Installation](#installation)
- [Dependencies](#dependencies)
- [FAQ](#faq)

# Project description

Sample application built with Next.js showcasing the [@api.video/media-stream-composer](https://github.com/apivideo/api.video-typescript-media-stream-composer) library.

The application is deployed on Heroku and can be viewed [here](https://record.a.video).

# Getting started

## Installation

To start the application:

- Edit the `.env` file to add your api.video upload token:

```bash
NEXT_PUBLIC_UPLOAD_TOKEN=[YOUR_API_VIDEO_UPLOAD_TOKEN] 
```

- Install dependencies and run the project:
  
```bash
npm i
npm run dev
```

- Then, open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

# Dependencies

We are using external library

| Plugin | README |
| ------ | ------ |
| [dotenv](https://www.npmjs.com/package/dotenv) | [Link to README](https://github.com/motdotla/dotenv) |
| [@mui/icons-material](https://www.npmjs.com/package/@mui/icons-material) | [Link to README](https://github.com/mui/material-ui) |
| [material-ui-popup-state](https://www.npmjs.com/package/material-ui-popup-state) | [Link to README](https://github.com/jcoreio/material-ui-popup-state) |
| [react-beautiful-dnd](https://www.npmjs.com/package/react-beautiful-dnd) | [Link to README](https://github.com/atlassian/react-beautiful-dnd#readme) |
| [react-color](https://www.npmjs.com/package/react-color) | [Link to README](https://github.com/casesandberg/react-color) |

# FAQ

If you have any questions, ask us in the [community](https://community.api.video). Or
use [issues](https://github.com/apivideo/private.a.video/issues).



CHANGELOG.md
# Changelog

All changes to this project will be documented in this file.

## [1.0.0] - 2023-03-13

- Application revamp
- New GitHub repository


src/nodeindex.js
require('dotenv').config();
const express = require('express');
const app = express();
var server = require('http').createServer(app);
app.use(express.static('public'));

var io = require('socket.io')(server);
var spawn = require('child_process').spawn;
var port =3035;


app.get('/', function(req, res,next) {  
    res.sendFile(__dirname + '/index1.html');
});

spawn('ffmpeg',['-h']).on('error',function(m){

	console.error("FFMpeg not found in system cli; please install ffmpeg properly or make a softlink to ./!");
	process.exit(-1);
});



io.on('connection', function(socket){
	console.log("connection");
    socket.emit('message','Hello from mediarecorder-to-rtmp server!');
	socket.emit('message','Please set rtmp destination before start streaming.');
	
	var ffmpeg_process, feedStream=false;
	//socket._vcodec='libvpx';//from firefox default encode
 

    socket.on('config_rtmpDestination',function(m){
		if(typeof m != 'string'){
			socket.emit('fatal','rtmp destination setup error.');
			return;
		}
		var regexValidator=/^rtmp:\/\/[^\s]*$/;//TODO: should read config
		if(!regexValidator.test(m)){
			socket.emit('fatal','rtmp address rejected.');
			return;
		}
		socket._rtmpDestination=m;
		socket.emit('message','rtmp destination set to:'+m);
        console.log('message','rtmp destination set to:'+m);
	}); 
   // console.log("rtmp: ", rtmpEndpoint);
    socket.on('config_vcodec',function(m){
		if(typeof m != 'string'){
			socket.emit('fatal','input codec setup error.');
			return;
		}
		if(!/^[0-9a-z]{2,}$/.test(m)){
			socket.emit('fatal','input codec contains illegal character?.');
			return;
		}//for safety
		socket._vcodec=m;
        console.group("socket._vcodec",socket._vcodec);
	}); 	

    //console.log('socket',socket);



	socket.on('start',function(m){
        
		if(ffmpeg_process || feedStream){
			
			socket.emit('fatal','stream already started.');
			console.log('fatal','stream already started.');
			return;
		}
		if(!socket._rtmpDestination){
			socket.emit('fatal','no destination given.');
			console.log('fatal','no destination given.');
			return;
		}
		var framerate = 25;
		
		console.log("apivideo server", socket._rtmpDestination);
		//default keyint is 250. but we want 2 for 1fps
		var key = Math.min(250, framerate*2);
		var keyint = "keyint="+key;
		//keyint_min default is 25
		var keyint_min = Math.min(25, framerate*2);
			var ops = [
				'-i','-',
				 //'-c:v', 'libx264', 
                 '-preset', 'ultrafast', '-tune', 'zerolatency', 
				//'-max_muxing_queue_size', '1000', 
				//'-bufsize', '5000',
				//'-r', '1', '-g', '2', '-keyint_min','2', 
			       '-r', framerate, '-g', framerate*2, '-keyint_min',keyint_min, 
				//	'-x264opts',keyint, '-crf', '25', '-pix_fmt', 'yuv420p',
			    //    '-profile:v', 'baseline', '-level', '3', 
     				'-c:a', 'aac', '-b:a','44k', '-ar', 44100, 
			        '-f', 'flv', socket._rtmpDestination		
		
		];
	
	    console.log("ops", ops);
		console.log("rtmp endpoint", socket._rtmpDestination);
		ffmpeg_process=spawn('ffmpeg', ops);
		console.log("ffmpeg spawned");
		feedStream=function(data){
			
			ffmpeg_process.stdin.write(data);
			//write exception cannot be caught here.	
		}

		ffmpeg_process.stderr.on('data',function(d){
			socket.emit('ffmpeg_stderr',''+d);
			console.log('ffmpeg_stderr',''+d);
		});
		ffmpeg_process.on('error',function(e){
			console.log('child process error'+e);
			socket.emit('fatal','ffmpeg error!'+e);
			feedStream=false;
			socket.disconnect();
		});
		ffmpeg_process.on('exit',function(e){
			console.log('child process exit'+e);
			socket.emit('fatal','ffmpeg exit!'+e);
			socket.disconnect();
		});
	});

	socket.on('binarystream',function(m){
		if(!feedStream){
			socket.emit('fatal','rtmp not set yet.');
			console.log('fatal','rtmp not set yet.');
			ffmpeg_process.stdin.end();
			ffmpeg_process.kill('SIGINT');
			return;
		}
		feedStream(m);
	});
	socket.on('disconnect', function () {
		console.log("socket disconected!");
		feedStream=false;
		if(ffmpeg_process)
		try{
			ffmpeg_process.stdin.end();
			ffmpeg_process.kill('SIGINT');
			console.log("ffmpeg process ended!");
		}catch(e){console.warn('killing ffmpeg process attempt failed...');}
	});
	socket.on('error',function(e){
		
		console.log('socket.io error:'+e);
	});
});

io.on('error',function(e){
	console.log('socket.io error:'+e);
});

server.listen(port, () =>
  console.log('Example app listening on port '+port+'!'),
);

public/index.js
RTMP_url = 'rtmp://broadcast.api.video/s/cd339b65-f4f9-40bc-afae-d586b3084f44';
live_url = 'https://embed.api.video/live/liqHiqZ8KXIW63tLHHAbRDL';
delegated_token = 'tompBU8ZJcUCEsKGxxoBydh';
//page defaults
//vod by default, but we can make the page default to live.
live = false;
//by default, screensahring
cameraOnly = false;
//adding a comment to change the doc for github

window.onload = function () {
    //this turns on the camera for a second - just to get permissions to populate the form with mics and cameras
    //navigator.getUserMedia = (navigator.mediaDevices.getUserMedia);
    navigator.mediaDevices.getUserMedia({ audio: true, video: true })
        .then((stream) => {
            getCamAndMics().then(() => stream.getTracks().forEach(x => x.stop()));
        })
        .catch(e => console.log(e));



    // is this a mobile device - no screen share - and 2 cameras?
    //see if screen capture is supported
    if ("getDisplayMedia" in navigator.mediaDevices) {
        console.log("screen capture supported");
    } else {
        console.log("screen capture NOT supported");
        cameraOnly = true;
    }

    //but we can change based on URL params
    const queryString = window.location.search;
    const urlParams = new URLSearchParams(queryString);
    const liveURLParam = urlParams.get('live');
    const cameraOnlyParam = urlParams.get('cameraOnly');
    if (liveURLParam === "true") {
        live = true;
        console.log("live set from url", live);
    }
    if (cameraOnlyParam === "true") {
        cameraOnly = true;
        console.log("cameraOnly set from url", cameraOnly);
        document.getElementById("screenpicker-select").options[0].disabled = true;
        document.getElementById("screenpicker-select").options[1].disabled = true;
        document.getElementById("screenpicker-select").options[2].disabled = true;
    }


    //set all the variables for the canvas and all the elements
    videoElem = document.getElementById("display");
    cameraElem = document.getElementById("camera");
    startElem = document.getElementById("start");
    stopElem = document.getElementById("stop");



    captionRecord = true;
    screenShared = false;
    //camera
    cameraW = 1280;
    cameraH = 720;
    cameraFR = 30;
    //set up the recording canvas
    canvas = document.getElementById("videoCanvas");

    ctx = canvas.getContext("2d");

    if (cameraOnly) {
        //record in portrait
        console.log("portrait canvas");
        ctx.canvas.width = 1280;
        ctx.canvas.height = 720;
        cameraW = 720;
        cameraH = 1280;

    } else {
        ctx.canvas.width = 1280;
        ctx.canvas.height = 720;
    }
    cw = ctx.width;
    ch = ctx.height;
    //caption font and colour
    ctx.font = "60px Arial";
    ctx.fillStyle = "red";
    ctx.textAlign = "center";


    //set xy coordinates for the screen and cameras
    screenX0 = 0;
    screenY0 = 0;
    screenX1 = ctx.canvas.width;
    screenY1 = ctx.canvas.height;

    cameraX0 = 0;
    cameraY0 = 0;
    cameraX1 = ctx.canvas.width / 3;
    cameraY1 = ctx.canvas.height / 3;



    //size of the video view whre the screen is stored
    screenWidth = ctx.canvas.width;
    screenHeight = ctx.canvas.height;
    //set dimensions for captions 
    var captionX = screenWidth / 2;
    var captionY = 100;
    ctx.clearRect(0, 0, canvas.width, canvas.height);


    //captionning
    interim_transcript = '';

    if ('webkitSpeechRecognition' in window) {
        console.log("speech recognition supported");
        recognition = new webkitSpeechRecognition();
    } else {
        console.log("speech recognition not supported");
        captionRecord = false;
    }
    console.log("captionRecord", captionRecord);

    //initialize captioning
    captioning();

    createStream();


    //  click listeners for 'tabs'
    liveButtom = document.getElementById('live');
    vodButton = document.getElementById('vodDiv');
    //set the initial state
    if (!live) {
        vodButton.className = "vodActive";
    } else {
        initializeLiveStream();
        liveButtom.className = "liveActive";
    }
    liveButtom.addEventListener('click', function () {
        initializeLiveStream();

    });
    vodButton.addEventListener('click', function () {
        document.getElementById("vod").checked = "true";
        console.log("checking VOD radio");
        live = false;
        vodButton.className = "vodActive";
        liveButtom.className = "live";
    });

};

function initializeLiveStream() {
    document.getElementById("liveStream").checked = "true";
    console.log("checking live stream radio");
    live = true;
    vodButton.className = "vod";
    liveButtom.className = "liveActive";
    connect_server();

}



function getCamAndMics() {
    // List cameras and microphones. in the menu
    var audioSelect = document.getElementById("audioPicker-select");
    var cameraSelect = document.getElementById("cameraPicker-select");

    return navigator.mediaDevices.enumerateDevices()
        .then(function (devices) {
            audioSelect.innerHTML = "";
            cameraSelect.innerHTML = "";
            devices.forEach(function (device) {
                console.log("device", device);
                deviceName = device.label;
                if (deviceName == "") {
                    deviceName = device.deviceId;
                }
                console.log(device.kind + ": named: " + deviceName + " id = " + device.deviceId);
                if (device.kind == "audioinput") {
                    //add a select to the audio dropdown list
                    var option = document.createElement("option");
                    option.value = device.deviceId;
                    option.text = deviceName;
                    audioSelect.appendChild(option);
                } else if (device.kind == "videoinput") {
                    //add a select to the camera dropdown list
                    var option = document.createElement("option");
                    option.value = device.deviceId;
                    option.text = deviceName;
                    cameraSelect.appendChild(option);

                }
            });
        })
        .catch(function (err) {
            console.log(err.name + ": " + err.message);
        });


}

function captioning() {


    //captioning function
    //captionning
    interim_transcript = '';
    var final_transcript = '';
    var recognizing = false;
    var ignore_onend;
    var start_timestamp;
    if (!captionRecord) {
        console.log("captions not supported in this browser!");
    } else {

        console.log("setting caption recognitions");
        recognition.continuous = true;
        recognition.interimResults = true;

        recognition.onstart = function () {
            recognizing = true;
        };

        recognition.onerror = function (event) {
            console.log("there was a caotioing error");
        };

        recognition.onend = function () {
            console.log("captioning stopped");
            recognizing = false;

        };

        recognition.onresult = function (event) {
            //heres where I'd put where stuff goes in my app....

            for (var i = event.resultIndex; i < event.results.length; ++i) {
                if (event.results[i].isFinal) {
                    interim_transcript = "";
                } else {
                    //append the words
                    interim_transcript = event.results[i][0].transcript;
                    console.log(interim_transcript);
                }
            }
        };
    }
}

function createStream() {
    //now lets capture the stream:
    var mediaSource = new MediaSource();

    var mediaRecorder;
    var recordedBlobs;
    var sourceBuffer;
    //capture stream at 25 fps
    stream = canvas.captureStream(35);
    console.log("stream tracks", stream.getTracks());
    console.log("stream tracks", stream.getVideoTracks());
    console.log("stream tracks", stream.getAudioTracks());
    // console.log("stream", stream);

    console.log("stream", stream);
    console.log('Got stream from canvas');
    var options = { mimeType: 'video/webm;codecs=vp9', bitsPerSecond: 100000 };

    //once the cameras and viewas are picked this will draw them on th canvas
    //the timeout is for 20ms, so ~50 FPS updates on the canvas
    function drawCanvas(screenIn, cameraIn, canvas) {
        var textLength = 60;
        canvas.drawImage(screenIn, screenX0, screenY0, screenX1, screenY1);
        canvas.drawImage(cameraIn, cameraX0, cameraY0, cameraX1, cameraY1);
        //write transcript on the screen
        if (interim_transcript.length < textLength) {
            ctx.fillText(interim_transcript, captionX, captionY);
        }
        else {
            ctx.fillText("no captions", captionX, captionY);

        }
        setTimeout(drawCanvas, 15, screenIn, cameraIn, canvas);

    }



    videoElem.addEventListener('play', function () {
        console.log('video playing');
        //draw the 2 streams to the canvas
        drawCanvas(videoElem, cameraElem, ctx);
    }, false);

    // Set event listeners for the start and stop buttons
    startElem.addEventListener("click", function (evt) {
        startCapture();
    }, false);

    stopElem.addEventListener("click", function (evt) {
        stopCapture();

    }, false);
}
async function startCapture() {
    try {
        //change buttons
        if (live) {
            stopElem.className = "stopLive";
            stopElem.style.display = "inline";
            startElem.style.display = "none";
        }
        else {
            stopElem.style.display = "inline";
            startElem.className = "modifyScreens";
            startElem.innerHTML = "Change layout";
        }

        //cedric wants to hide the canvas
        canvasShow = document.getElementById("canvasDisplay-checkbox").checked;
        console.log("canvasShow", canvasShow);
        //hide the canvas if canvasShow is not checked
        if (!canvasShow) {
            //hide the canvas
            canvas.style.display = "none";
        }


        //add text for the 2 screens
        document.getElementById("videoInputText").innerHTML = "Screen and camera inputs";

        //arrange the cameras/screens with CSS
        var screenLayout = document.getElementById("screenpicker-select").value;
        if (cameraOnly) {
            screenLayout = 'cameraOnly';
        }
        console.log("screenLayout", screenLayout);

        if (screenLayout === 'screenOnly') {
            //no camera
            //big screen
            screenX0 = 0;
            screenY0 = 0;
            screenX1 = ctx.canvas.width;
            screenY1 = ctx.canvas.height;

            cameraX0 = 0;
            cameraY0 = 0;
            cameraX1 = 0;
            cameraY1 = 0;

        } else if (screenLayout === 'cameraOnly') {
            //big camera 
            //no screen
            screenX0 = 0;
            screenY0 = 0;
            screenX1 = 0;
            screenY1 = 0;

            cameraX0 = 0;
            cameraY0 = 0;
            cameraX1 = ctx.canvas.width;
            cameraY1 = ctx.canvas.height;

        } else if (screenLayout === 'bottomRight') {
            //bottom right camera
            //big screen
            screenX0 = 0;
            screenY0 = 0;
            screenX1 = ctx.canvas.width;
            screenY1 = ctx.canvas.height;

            cameraX0 = .625 * ctx.canvas.width;
            cameraY0 = .625 * ctx.canvas.height;
            cameraX1 = ctx.canvas.width / 3;
            cameraY1 = ctx.canvas.height / 3;

        } else {
            //default bottom left camera
            //big screen
            screenX0 = 0;
            screenY0 = 0;
            screenX1 = ctx.canvas.width;
            screenY1 = ctx.canvas.height;

            cameraX0 = 20;
            cameraY0 = .625 * ctx.canvas.height;
            cameraX1 = ctx.canvas.width / 3;
            cameraY1 = ctx.canvas.height / 3;

        }

        //where do captions go?
        var captionLocation = document.getElementById("captionspicker-select").value;
        captionX = 0;
        captionY = 0;
        if (!captionRecord) {
            //if captions are not supported - turn 'em off
            captionLocation = "noCaptions";
        }
        console.log("captions", captionLocation);
        if (captionLocation === "captionsTop") {
            //captions at the top
            captionX = screenWidth / 2;
            captionY = 50;
            startCaptions();

        } else if (captionLocation === "captionsBottom") {
            //captions at bottom
            captionX = screenWidth / 2;
            //this will allow for 3 lines to appear
            captionY = screenHeight - 90;
            startCaptions();
        }
        //if no captions, we dont start captions.

        function startCaptions() {
            //start captioning

            final_transcript = '';
            //english, but change as desired.
            //recognition.lang = "en-GB";
            recognition.start();
        }

        //rebuild the screen options
        //only thing is muting the audio
        //this prevents awful feedback..
        displayMediaOptions = {
            video: {
                frameRate: { ideal: cameraFR }
            },
            audio: false
        };
        console.log(JSON.stringify(displayMediaOptions));



        //select the camera and the micrphone: 
        var cameras = document.getElementById("cameraPicker-select");
        var cameraId = cameras.options[cameras.selectedIndex].value;
        var mics = document.getElementById("audioPicker-select");
        var micId = mics.options[mics.selectedIndex].value;

        //camera
        console.log("frameRate", cameraFR);
        navigator.getUserMedia = (navigator.mediaDevices.getUserMedia ||
            navigator.mediaDevices.mozGetUserMedia ||
            navigator.mediaDevices.msGetUserMedia ||
            navigator.mediaDevices.webkitGetUserMedia);
        var videoOptions = {
            deviceId: cameraId,
            width: { min: 100, ideal: cameraW, max: 1920 },
            height: { min: 100, ideal: cameraH, max: 1080 },
            frameRate: { ideal: cameraFR }
        };
        if (cameraOnly) {
            videoOptions = {
                facingMode: "user",
                width: { min: 100, ideal: cameraW, max: 1920 },
                height: { min: 100, ideal: cameraH, max: 1080 },
                frameRate: { ideal: cameraFR }
            };
        }
        cameraMediaOptions = {
            audio: false,
            video: videoOptions
        };
        rearCameraMediaOptions = {
            audio: false,
            video: {
                facingMode: "environment",
                width: { min: 100, ideal: cameraW, max: 1920 },
                height: { min: 100, ideal: cameraH, max: 1080 }
            }
        };
        console.log(JSON.stringify(cameraMediaOptions));

        //all settings in  - start the cameras screenshare
        if (!screenShared) {

            if (!cameraOnly) {
                //share the screen - it has not been shared yet!
                videoElem.srcObject = await navigator.mediaDevices.getDisplayMedia(displayMediaOptions);
                screenShared = true;
            } else {
                //screen cannot be shared, so grab the front facing camera.
                videoElem.srcObject = await navigator.mediaDevices.getUserMedia(rearCameraMediaOptions);
                screenShared = true;

            }
        }
        cameraStream = await navigator.mediaDevices.getUserMedia(cameraMediaOptions);
        cameraElem.srcObject = cameraStream;

        audioStreamOptions = {
            mimeType: "video/webm;codecs=vp8,opus",
            // mimeType: "video/mp4",
            audio: {
                deviceId: micId
            }
        };




        //grab the audio and add it to the stream coming from the canvas
        audioStream = await navigator.mediaDevices.getUserMedia(audioStreamOptions);


        for (const track of audioStream.getTracks()) {
            console.log("adding audio track");
            stream.addTrack(track);
            console.log("stream added audio", stream);
        }

        console.log("stream tracks", stream.getTracks());
        console.log("stream tracks", stream.getVideoTracks());
        console.log("stream tracks", stream.getAudioTracks());

        //JUST START RECORDING
        startRecording(live);

    } catch (err) {
        console.error("Error: " + err);
    }
}
function stopCapture(evt) {
    //FIX BUTTONS
    //change buttons
    if (live) {

        startElem.className = "start";
        startElem.innerHTML = "Reload to stream again";
        startElem.disabled = true;
        socket.close();
    } else {
        startElem.className = "start";
        startElem.innerHTML = "Start";
    }
    startElem.style.display = "inline";
    stopElem.style.display = "none";
    document.getElementById("videoInputText").innerHTML = "";
    //screen stop
    stopRecording();
    let tracks = videoElem.srcObject.getTracks();
    console.log(JSON.stringify(tracks));
    tracks.forEach(track => track.stop());
    videoElem.srcObject = null;
    screenShared = false;

    //camera stop
    let cameraTracks = cameraElem.srcObject.getTracks();
    console.log(JSON.stringify(tracks));
    tracks.forEach(cameraTracks => cameraTracks.stop());
    cameraElem.srcObject = null;

    //captions stop
    if (captionRecord) {
        recognition.stop();
    }
    if (!live) {
        //stop blob recording
        //uploadTheVideo();
        // download();
    }


}

function startRecording(live) {
    //if I omit the MIMEtype, MediaRecorder works in Safari 14.0.3.  If I add a Mime.... it fails.
    //i had a mimetype in the options and it would not record properly.
    var options = { audioBitsPerSecond: 100000, videoBitsPerSecond: 4000000 };
    //var options = {};
    // var options = 'video/webm';
    recordedBlobs = [];
    try {
        if (live) {
            mediaRecorder = new MediaRecorder(stream, options);
        } else {
            //instatiate the apivideo meida recorder upload function
            console.log("create apivideo media recorder!");
            mediaRecorder = new ApiVideoMediaRecorder(stream, {
                uploadToken: delegated_token
            });
        }
        console.log("options", options);
        console.log("mediaRecorder mime", mediaRecorder.mimeType);
    } catch (e2) {
        alert('MediaRecorder is not supported by this browser.');
        console.error('Exception while creating MediaRecorder:', e2);
        return;
    }
    console.log('Created video MediaRecorder', mediaRecorder, 'with options', options);
    console.log(",ediacrecorder stream info", mediaRecorder.stream);
    // console.log(",ediacrecorder stream trackinfo", mediaRecorder.stream.getTracks());
    mediaRecorder.onstop = handleStop;
    if (live) {
        console.log("live mime", mediaRecorder.mimeType);
        socket.emit("config_vcodec", mediaRecorder.mimeType);
        mediaRecorder.ondataavailable = function (e) {
            // console.log("e", e.data);
            socket.emit("binarystream", e.data);
            state = "start";
            //chunks.push(e.data);
        }
        document.getElementById("video-information").innerHTML = "Live Stream available after 20s <a href=" + live_url + ">here</a>";
    }
    mediaRecorder.start(10); // collect 10ms of data
    console.log('MediaRecorder started', mediaRecorder);
}

function handleStop(event) {
    console.log('Recorder stopped: ', event);
    console.log('Recorded Blobs: ', recordedBlobs);
}

function stopRecording() {


    document.getElementById("video-information").innerHTML = "Uploading the last bit of the video.  Please wait a second."
    mediaRecorder.stop()//;
        .then(function (video) {

            console.log(video);
            //the video is fully uploaded. there will now be a url in the response
            playerUrl = video.assets.player;
            console.log("all uploaded! Watch here: ", playerUrl);
            document.getElementById("video-information").innerHTML = "all uploaded! Watch the video <a href=\'" + playerUrl + "\' target=\'_blank\'>here</a>";

        });

    //  recordedVideo.controls = true;
    // download();
}

function play() {
    var type = (recordedBlobs[0] || {}).type;
    var superBuffer = new Blob(recordedBlobs, { type });
    recordedVideo.src = window.URL.createObjectURL(superBuffer);
}

function download() {
    var blob = new Blob(recordedBlobs, { type: 'video/webm' });
    var url = window.URL.createObjectURL(blob);
    var a = document.createElement('a');
    a.style.display = 'none';
    a.href = url;
    a.download = 'test.webm';
    document.body.appendChild(a);
    a.click();
    setTimeout(function () {
        document.body.removeChild(a);
        window.URL.revokeObjectURL(url);
    }, 100);
}

function connect_server() {

    if (!stream) { fail('No getUserMedia() available.'); }
    if (!MediaRecorder) { fail('No MediaRecorder available.'); }


    var socketOptions = { secure: true, reconnection: true, reconnectionDelay: 1000, timeout: 10000, pingTimeout: 10000, pingInterval: 2000 };

    //start socket connection
    socket = io.connect("/", socketOptions);
    // console.log("ping interval =", socket.pingInterval, " ping TimeOut" = socket.pingTimeout);
    //output_message.innerHTML=socket;

    socket.on('connect_timeout', (timeout) => {
        console.log("state on connection timeout= " + timeout);
        // output_message.innerHTML="Connection timed out";
        // recordingCircle.style.fill='gray';

    });
    socket.on('error', (error) => {
        console.log("state on connection error= " + error);
        //   output_message.innerHTML="Connection error";
        //    recordingCircle.style.fill='gray';
    });

    socket.on('connect_error', function () {
        console.log("state on connection error= " + state);
        //   output_message.innerHTML="Connection Failed";
        //   recordingCircle.style.fill='gray';
    });

    socket.on('message', function (m) {
        console.log("state on message= " + state);
        console.log('recv server message', m);
        //  show_output('SERVER:'+m);

    });

    socket.on('fatal', function (m) {

        // show_output('Fatal ERROR: unexpected:'+m);
        //alert('Error:'+m);
        console.log("fatal socket error!!", m);
        console.log("state on fatal error= " + state);
        //already stopped and inactive
        console.log('media recorder restarted');
        // recordingCircle.style.fill='gray';



    });

    socket.on('ffmpeg_stderr', function (m) {
        //this is the ffmpeg output for each frame
        console.log('FFMPEG:' + m);


    });

    socket.on('disconnect', function (reason) {
        console.log("state disconec= " + state);
        // show_output('ERROR: server disconnected!');
        console.log('ERROR: server disconnected!' + reason);
        //  recordingCircle.style.fill='gray';
        //reconnect the server
        // connect_server();



    });

    state = "ready";
    console.log("state = " + state);
    console.log('config_rtmpDestination', RTMP_url);
    socket.emit('config_rtmpDestination', RTMP_url);
    socket.emit('start', 'start');

}
